homeassistant/components/icloud/device_tracker.py

```diff
@@ -1,547 +1,119 @@
-            except PyiCloudException as error:
-                _LOGGER.error("Error setting up 2FA: %s", error)
-        else:
-            self.api.authenticate()
-
-        currentminutes = dt_util.now().hour * 60 + dt_util.now().minute
-        try:
-            for devicename in self.devices:
-                interval = self._intervals.get(devicename, 1)
-                if (currentminutes % interval == 0) or (
-                    interval > 10 and currentminutes % interval in [2, 4]
-                ):
-                    self.update_device(devicename)
-        except ValueError:
-            _LOGGER.debug("iCloud API returned an error")
-
-    def determine_interval(self, devicename, latitude, longitude, battery):
-        """Calculate new interval."""
-        currentzone = run_callback_threadsafe(
-            self.hass.loop, async_active_zone, self.hass, latitude, longitude
-        ).result()
-
-        if (
-            currentzone is not None
-            and currentzone == self._overridestates.get(devicename)
-        ) or (currentzone is None and self._overridestates.get(devicename) == "away"):
-            return
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
 
-        zones = (
-            self.hass.states.get(entity_id)
-            for entity_id in sorted(self.hass.states.entity_ids("zone"))
+    @property
+    def battery_level(self):
+        """Return the battery level of the device."""
+        return self._device.battery_level
+
+    @property
+    def source_type(self):
+        """Return the source type, eg gps or router, of the device."""
+        return SOURCE_TYPE_GPS
+
+    @property
+    def icon(self):
+        """Return the icon."""
+        return icon_for_icloud_device(self._device)
+
+    @property
+    def device_state_attributes(self):
+        """Return the device state attributes."""
+        return self._device.state_attributes
+
+    async def async_added_to_hass(self):
+        """Register state update callback."""
+        self._unsub_dispatcher = async_dispatcher_connect(
+            self.hass, TRACKER_UPDATE, self._async_receive_data
         )
 
-        distances = []
-        for zone_state in zones:
-            zone_state_lat = zone_state.attributes["latitude"]
-            zone_state_long = zone_state.attributes["longitude"]
-            zone_distance = distance(
-                latitude, longitude, zone_state_lat, zone_state_long
-            )
-            distances.append(round(zone_distance / 1000, 1))
-
-        if distances:
-            mindistance = min(distances)
-        else:
-            mindistance = None
+    async def async_will_remove_from_hass(self):
+        """Clean up after entity before removal."""
+        self._unsub_dispatcher()
 
-        self._overridestates[devicename] = None
+    @callback
+    def _async_receive_data(self):
+        """Update device data."""
+        self.async_write_ha_state()
```

Comment:
If we don't need to do anything more here, we can connect the signal to `self.async_write_ha_state` directly.